pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -575,7 +575,13 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 		// need to restart it.
 		if containerStatus == nil || containerStatus.State != kubecontainer.ContainerStateRunning {
 			if kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus) {
-				message := fmt.Sprintf("Container %+v is dead, but RestartPolicy says that we should restart it.", container)
+				var message string
+				if containerStatus == nil {
+					message = fmt.Sprintf("Starting container: %+v", container)
```

Comment:
I think its fair to say that precision around this message is hard, and so the question is if varying the log message helps provide additional clarity at this point in the call path.  if all historical containers were garbage collected, no status would be known.  if the container status is unknown, or in created state, we would need to restart it as well.

Maybe a better message is the following:

"Container %q of pod %q is not in desired state and shall be started".

The `ShouldContainerBeRestarted` method has an opportunity to provide more detailed log message for the container is not in a desired state, and in some cases already has logging to provide that message.

Reply:
@derekwaynecarr thanks! Good point. Do you think it is okay, then, to:
 * Add the line you suggested in case the status is not nil
 * Add the "Starting..." line in case it is nil (if it's nil, it was never started)
 * And, not sure, maybe also add a log with higher verbosity (`klog.V(4)`) in `ShouldRestartContainer()` for this case: https://github.com/kubernetes/kubernetes/blob/628102d03820f4213b77267699aba5f649ea2f9f/pkg/kubelet/container/helpers.go#L80

IMHO, in this part of the code path we are commenting on, as is executed to start regular containers on pods that have initContainers (as I explained in the commit msg), logging here with V(3) seems useful. Another alternative can be logging all on `ShouldRestartContainer()` cases and not here, but that seems a little excessive, specially if some other user is added in the future. I'd keep the logging inside that function with high verbosity for the "weird" cases, as it is know. And keep the message here with `klog.V(3)`.

All the other "wierd" cases are logged in `ShouldContainerBeRestarted()`, this is the only "weird" case not logged.

The other usage of `ShouldContainerBeRestarted()` is in `kubelet_pods.go`, `convertToAPIContainerStatuses()` that goes over failed contaiers and checks. So, adding a log for the "weird" case in `ShouldContainerBeRestarted()` with V(4) level seems fine there as well.

What do you think? Does it make sense?

Reply:
I guess I don't really understand the motivation for this log message change.

I still prefer:
"Container %q of pod %q is not in desired state and shall be started".

Kubelet is just a reconciling state machine, and all we care about is the container is not as desired....

Reply:
The motivation comes from looking at the logs and see that a container was dead, when in fact it wasn't and trying to understand what was wrong. I found that nothing was wrong, the log was just bogus.

I'll change the message to say just that, if it looks okay for you too :-)

Thanks!